File: bindings/cpp/test/test_pipeline.cpp
// Pipeline — remote-only (contract §6 "Pipeline").
#include "harness.hpp"
#include "kevy/client.hpp"
#include "kevy/pipeline.hpp"

using namespace kevy;
using namespace kevy::test;

KEVY_TEST(pipeline_order_and_inline_errors) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  auto replies = c.pipeline([](PipelineBuf& p) {
    p.cmd({"SET", "k", "v"});
    p.cmd({"INCR", "k"});   // -ERR (k is not an integer) — must land INLINE
    p.cmd({"GET", "k"});
  });
  CHECK_EQ(replies.size(), size_t(3));
  CHECK(replies[0].kind == ReplyKind::Simple && replies[0].str() == "OK");
  CHECK(replies[1].is_error());  // inline error, batch not aborted
  CHECK(replies[2].kind == ReplyKind::Bulk && replies[2].str() == "v");
}

KEVY_TEST(pipeline_empty_no_wire) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  auto replies = c.pipeline([](PipelineBuf&) {});  // empty → [] without wire I/O
  CHECK_EQ(replies.size(), size_t(0));
}

KEVY_TEST(pipeline_empty_argv_invalid) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  CHECK_THROWS_KIND(c.pipeline([](PipelineBuf& p) {
    p.cmd({"PING"});
    p.cmd({});  // empty argv poisons the batch
  }), ErrorKind::InvalidInput);
}

File: bindings/python/tests/test_pipeline.py
"""Pipeline (non-atomic batching, contract §3.13, §6). Remote-only."""

import kevy
from kevy import ReplyKind


def test_pipeline_order_and_inline_errors(remote_server):
    c = kevy.connect(remote_server.url)
    replies = c.pipeline(
        lambda p: p.cmd(b"SET", b"k", b"v")
        .cmd(b"GET", b"k")
        .cmd(b"INCR", b"k")  # k is not an integer → inline error, batch not aborted
        .cmd(b"GET", b"k")
    )
    assert len(replies) == 4
    assert replies[0].data == b"OK"
    assert replies[1].data == b"v"
    assert replies[2].is_error()  # per-command error, inline (did not abort)
    assert replies[3].data == b"v"
    c.close()


def test_empty_batch_no_io(remote_server):
    c = kevy.connect(remote_server.url)
    assert c.pipeline(lambda p: None) == []
    c.close()


def test_empty_argv_poisons(remote_server):
    c = kevy.connect(remote_server.url)
    try:
        c.pipeline(lambda p: p.cmd(b"PING").cmd())  # empty argv
        assert False, "expected InvalidInputError"
    except kevy.InvalidInputError:
        pass
    c.close()


def test_pipeline_len(remote_server):
    c = kevy.connect(remote_server.url)

    def build(p):
        p.cmd(b"PING")
        p.cmd(b"PING")
        assert len(p) == 2 and not p.is_empty()

    c.pipeline(build)
    c.close()


def test_pipeline_embedded_unsupported():
    c = kevy.connect("mem://pipe")
    try:
        c.pipeline(lambda p: p.cmd(b"PING"))
        assert False
    except kevy.UnsupportedError:
        pass
    c.close()
